cpp/src/arrow/c/abi.h

```diff
@@ -0,0 +1,65 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define ARROW_FLAG_DICTIONARY_ORDERED 1
+#define ARROW_FLAG_NULLABLE 2
+#define ARROW_FLAG_MAP_KEYS_SORTED 4
+
+struct ArrowSchema {
+  // Array type description
+  const char* format;
+  const char* name;
+  const char* metadata;
+  int64_t flags;
+  int64_t n_children;
+  struct ArrowSchema** children;
+  struct ArrowSchema* dictionary;
+
+  // Release callback
+  void (*release)(struct ArrowSchema*);
+  // Opaque producer-specific data
+  void* private_data;
+};
+
+struct ArrowArray {
+  // Array data description
+  int64_t length;
+  int64_t null_count;
+  int64_t offset;
```

Comment:
offset and n_buffers seems out of place here. What is the rationale for including these. n_buffers is known by definition so why add it here? offset seems like an optimization that could solved by shifting the pointers rather than providing an offset.

Reply:
Note, I understand that offset can allow one to avoid bit-shifting but that seems like something we shouldn't support or tell people to do.

Reply:
The cpp code supports slices of arrays/record-batches without doing a copy, by just tracking the offset (even for the validity buffers). shifting the pointers doesn't work for validity bitmaps.

Reply:
Yeah, without the offset member, zero-copy slicing is simply not possible. This would be a serious deficiency

Reply:
This is consciously not supported at the arrow format level (unless my memory fails me). I don't see why we should support it here. This addition would make this coupled to the cpp implementation (and incompatible with other implementation), which it shouldn't be.

Reply:
@jacques-n regarding the n_buffers, a couple things come to mind:

* In C I think it's generally considered a best practice to provide the length of variable-size heap-allocated arrays
* To obtain the length of "buffers" for each array / subarray by examining the schema could potentially be a nuisance as this data structure would no longer be "self-contained" (or at least, able to have its contents examined easily). For example, a function that creates a copy of the C structure would need the corresponding schema reference in order to know how large the buffers array is

Reply:
Re n_buffers: Since we're requiring the understanding of how to map logical length to physical length via schema (to know the amount of addressable memory), how is this different? In either circumstances, you'll be unable to work with the object unless you know schema.

Reply:
Right, but consider as a straw man the implementation of a function that creates a copy of a struct

```
void CopyTo(struct ArrowArray* src, struct ArrowArray* dst) {
  ...
}
```

The fact of having to query the schema to know how large the buffers array is within each level of `src` definitely seems like a nuisance to me. For a C structure to lack knowledge of the size of any of its heap-allocated array members seems to go against my intuition

Reply:
I don't understand your comment...

ATM, one needs to know schema information to consume or copy buffers. At a minimum I need both how many buffers and how long each buffer is. Right now, the definition includes one but not the other. Why not exclude both (or include both)?

Reply:
I meant that it would not be possible to copy `src->buffers` (the array of memory addresses) to `dst->buffers` without querying the schema to determine how long the array is supposed to be. 

It does seem useful to include the buffer sizes.

Reply:
Putting `n_buffers` here merely is meant as an implementation and diagnosis help. I agree it's strictly not required. But since this is gonna be produced or consumed by developers with less expertise of the Arrow memory format than us, I think it could help detect and avoid mistakes. Also, it's very cheap to produce and store.

(for example, if a producer gives you a string array with only two buffers, you immediately know it has badly implemented the spec - without having to investigate a pointer dereference crash)

I agree we may want to add buffer sizes for similar reasons. It would add more weight to the structure (*). I don't have a strong opinion.

(*) There are two possibilities:

```c
struct ArrowBuffer {
  const void* data;
  int64_t nbytes;
};

struct ArrowArray {
  // ...
  struct ArrowBuffer* buffers;
};
```
or:
```c

struct ArrowArray {
  // ...
  const void** buffers;
  int64_t buffer_sizes;
};
```

Reply:
Both of these feel very much like the [VectorLayout](https://github.com/apache/arrow/blob/e081a4c27a5a592251f9f325a05479d4120e30e6/format/Message.fbs#L113) stuff in the early ipc spec. It was meant for better message validation but ultimately was just generated using lookup tables in most impls. I suppose this is slightly better since it is assumed users won't have anything but this header but I'm inclined not to do any of this kind of "validation encoding".

Reply:
@wesm, I also maintain that offset does not belong here (I didn't see a response to my previous comment on this point). It is an internal concept of the c++ library and we shouldn't tie this interface definition to that library. This should be a subset of the core format definition in terms of schema and data structures, not a superset. (The release operation is a functional definition, not a data structure.)

Reply:
> It is an internal concept of the c++ library and we shouldn't tie this interface definition to that library. 

I don't agree with this. Any library that implements zero copy slicing must use the "offset" concept. This is used in C++, C#, Rust, Go, and JavaScript. In fact, Java is the **only** language that does not support it. 

Without this, any library that has implemented slicing will necessitate wasteful copying to be able to use this data structure. Since we can't change the struct after adopting the proposal, I don't see why we would want to cripple it in this way from the get-go

Reply:
It seems clear that we have different opinions about whether this pattern should be supported. I'm strongly against it for both technical reasons and because it isn't supported by the Arrow specification. I'm sure we came to our conclusions through different experiences. Let me provide a little background on my perspective.

The java codebase actually did have this concept at one point (well before it was part of the Arrow project) and we chose to eliminate it as we ultimately came to the conclusion that it made the wrong tradeoffs between efficient calculations and efficient slicing. What we discovered was that any time you were trying to do zero-copy slicing, doing so on an eight value boundary was a much better design choice than have the additional complexity to deal with bit-shifting on the computation side.

I see excluding this attribute/option as a prudent design choice that kernel developers will benefit from. In general, I expect there to be far more computation algorithms and kernels being written than slicing logic. The number of times that computational code becomes more complex on the processing side with the offset field seems to far outweigh any benefits of byte-splitting on the slicing side. 

It would be helpful to better understand why you feel this is so crippling? Which specific algorithms do you see impacted by disallowing byte-splitting? How common are these and what do you think the cost is of not supporting this. How does that compare with your sense of the cost of dealing with this in every computational algorithm?

Reply:
If this data interface were restricted to the domain of computational kernel evaluation, then I would agree with you, but there is a wider diversity of use cases that the project supports. Note that in Dremio the slice-and-copy pattern is less onerous because record batches are generally limited in size to 64K elements and less, so the calculus is different when record batches can be arbitrarily large in general. 

So, first we must consider it already as a given that most implementations are going to provide zero-copy slicing as a feature for in-memory data. The question then is how to handle slices with offset that is not a multiple of 8. There are 3 options:

* Disallow such slicing at the C interface boundary, forcing copy-materialization for all use cases
* Allow slicing, but regularize sliced bitmaps (and the data buffer for sliced BooleanArray) for all computational kernels, i.e. assume an offset of 0 for all kernels
* Allow slicing and implement kernels for non-zero offsets

The 2nd option seems like the most practical -- the interface is always zero copy and kernel developers don't bother trying to optimize for non-zero offsets. It does mean that some functions that are unable to act on sliced data must check whether the offset is 0 and if not, convert the input to be offset-0 based (allocating a new bitmap and copying over the bits if necessary)

I'm not going to block up the work over this but it would feel like a shame to me to have a C data interface which is not zero copy all of the time.

Reply:
> I don't think this would be tenable in C++

Why not? Where is the critical need for allowing non-regularized slicing? Requiring that anything that is slicing the data does so div-8 boundary seems like a pretty low-burden constraint. Are you seeing lots of algorithms where this is high burden? 

I'm not arguing here for copy materialization, I'm saying we should make non-div-8 slices invalid and have the ability for zero copy slices and computational simplicity.

Reply:
> Requiring that anything that is slicing the data does so div-8 boundary seems like a pretty low-burden constraint

I think the use-case is a user chooses an arbitrary slice.  E.g. I have  datatable, and I choose to slice rows 1 to N.  the library shouldn't throw an exception here, so it is up to implementors to either do the copy or keep the slice index.  But perhaps there is a third option?

Reply:
> Where is the critical need for allowing non-regularized slicing?

As background, I started my career in financial analytics where the world is more oriented around time series data where slicing and analytics on large memory-mapped data sets is the general expectation. Think application of rolling time series windows functions. Consider as an example such system oriented around this kind of data https://en.wikipedia.org/wiki/Kdb%2B

This kind of analytics is very different from batch processing SQL based workloads. Apache Arrow already has a lot of users from the financial analytics world and many of them are going to be using C++. If you start requiring memory copying every time you request an arbitrary window of data (big or small, a window could be 1KB or 1GB) from a large time series you cripple this use case.